src/tactic/unfold_cases.lean

```diff
@@ -0,0 +1,169 @@
+  followed by unfolding the definition:
+  ```lean
+  ⊢ ff.cases_on
+    (x.cases_on (id_rhs bool ff) (id_rhs bool ff))
+    (x.cases_on (id_rhs bool ff) (id_rhs bool tt)) = ff
+  ```
+
+  At this point, it can make progress using `dsimp`. But then it gets stuck:
+  ```lean
+  ⊢ bool.rec (id_rhs bool ff) (id_rhs bool ff) x = ff
+  ```
+
+  Next, it can introduce a case split on `x`. At this point, it has to prove two
+  goals:
+  ```lean
+  ⊢ bool.rec (id_rhs bool ff) (id_rhs bool ff) ff = ff
+  ⊢ bool.rec (id_rhs bool ff) (id_rhs bool ff) tt = ff
+  ```
+
+  Now, however, both goals can be discharged using `refl`.
+-/
+namespace tactic
+open expr
+namespace unfold_cases
+/--
+  Given an equation `f x = y`, this tactic tries to infer an expression that can be
+  used to do distinction by cases on to make progress.
+
+  Pre-condition: assumes that the outer-most application cannot be beta-reduced
+  (e.g. `whnf` or `dsimp`).
+-/
+meta def find_splitting_expr : expr → tactic expr
+| `(@ite %%cond %%dec_inst _ _ _ = _) := pure `(@decidable.em %%cond %%dec_inst)
+| `(%%l@(app x y) = _) := pure y
+| e := fail!"expected an expression of the form: f x = y. Got:\n{e}"
+
+/--
+  Tries to finish the current goal using the `inner` tactic. If the tactic
+  fails, it tries to find an expression on which to do a distinction by
+  cases and calls itself recursively.
+
+  The order of operations is significant. Because the unfolding can potentially
+  be infinite, it is important to apply the `inner` tactic at every step.
+
+  Notice, that if the `inner` tactic succeeds, the recursive unfolding is stopped.
+-/
+meta def unfold_cases_core (inner : interactive.itactic) : tactic unit :=
+inner <|>
+(do split_ifs [], all_goals unfold_cases_core, skip) <|>
+do
+  tgt ← target,
+  e ← find_splitting_expr tgt,
+  focus1 $ do
+    cases e,
+    all_goals $ (dsimp_target >> unfold_cases_core) <|> skip,
+    skip
+
+/--
+  Given a target of the form `⊢ f x₁ ... xₙ = y`, unfolds `f` using a delta reduction.
+-/
+meta def unfold_tgt : expr → tactic unit
+| e@`(%%l@(app _ _) = %%r) :=
+  match l.get_app_fn with
+  | const n ls := delta_target [n]
+  | e := fail!"couldn't unfold:\n{e}"
+  end
+| e := fail!"expected an expression of the form: f x = y. Got:\n{e}"
+end unfold_cases
+
+namespace interactive
+open unfold_cases
+
+/--
+  This tactic unfolds the definition of a function or `match` expression.
+  Then it recursively introduces a distinction by cases. The decision what expression
+  to do the distinction on is driven by the pattern matching expression.
+
+  A typical use case is using `unfold_cases { refl }` to collapse cases that need to be
+  considered in a pattern matching.
+
+  ```lean
+  have h : foo x = y, by unfold_cases { refl },
+  rw h,
+  ```
+
+  The tactic expects a goal in the form of an equation, possibly universally quantified.
+
+  Further examples can be found in `test/unfold_cases.lean`.
```

Comment:
Could you show one complete example in the doc string here? This will show up as the [doc entry](https://leanprover-community.github.io/mathlib_docs/tactics.html) for `unfold_cases`, and it's clearest to have an example visible there. (This isn't to say that every tactic has such an example. We're working on it.)

Reply:
sure, will do.